# Project - Natural Language Processing (NLP)

This Project aimes to:

1- Collect data from the Brazilian Monetary Policy Committee (COPOM).

2- Based on this data, predict the most probable trajectory outcome from the next meeting, regarding the basic interest rate of the Central Bank of Brazil (SELIC rate).

## Importing Libraries

In [264]:
import pandas as pd
pd.set_option("display.max_colwidth", 150)
pd.set_option("display.min_rows", 20)

import matplotlib
matplotlib.rcParams["figure.figsize"] = (36,16)

import urllib.request, json

from tqdm import tqdm

from bs4 import BeautifulSoup
import re

import nltk
from nltk.probability import FreqDist
from nltk.tokenize import word_tokenize
import string
import pdfplumber
import fitz  # PyMuPDF

import os
import glob

import matplotlib.pyplot as plt

## Collecting minutes of Monetary Policy Committee meetings, from Brazil's central bank

### Colection the total available and the numbers of each minute

In [265]:
# Using 1000 to ensure all available data is being collected
url = 'https://www.bcb.gov.br/api/servico/sitebcb/copom/atas?quantidade=1000'

In [266]:
with urllib.request.urlopen(url) as url:
    data = json.load(url)
    print(data)

{'conteudo': [{'nroReuniao': 260, 'dataReferencia': '2024-01-31', 'dataPublicacao': '2024-02-06', 'titulo': '260ª Reunião - 30-31 janeiro, 2024'}, {'nroReuniao': 259, 'dataReferencia': '2023-12-13', 'dataPublicacao': '2023-12-19', 'titulo': '259ª Reunião - 12-13 dezembro, 2023'}, {'nroReuniao': 258, 'dataReferencia': '2023-11-01', 'dataPublicacao': '2023-11-07', 'titulo': '258ª Reunião - 31 outubro - 1 novembro, 2023'}, {'nroReuniao': 257, 'dataReferencia': '2023-09-20', 'dataPublicacao': '2023-09-26', 'titulo': '257ª Reunião - 19-20 setembro, 2023'}, {'nroReuniao': 256, 'dataReferencia': '2023-08-02', 'dataPublicacao': '2023-08-08', 'titulo': '256ª Reunião - 1-2 agosto, 2023'}, {'nroReuniao': 255, 'dataReferencia': '2023-06-21', 'dataPublicacao': '2023-06-27', 'titulo': '255ª Reunião - 20-21 junho, 2023'}, {'nroReuniao': 254, 'dataReferencia': '2023-05-03', 'dataPublicacao': '2023-05-09', 'titulo': '254ª Reunião - 2-3 maio, 2023'}, {'nroReuniao': 253, 'dataReferencia': '2023-03-22', '

Converting the .json content into a dataframe

In [267]:
df_aux = pd.DataFrame(data['conteudo'])

Filtering only the meeting minutes from number 202 onwards.

The standardization of the communications was different before the mentioned period.

In [268]:
index_202 = df_aux[df_aux['nroReuniao'] == 202].index[0]
df_aux = df_aux.loc[:index_202]

In [269]:
df_aux

nroReuniao dataReferencia dataPublicacao  \
0          260     2024-01-31     2024-02-06   
1          259     2023-12-13     2023-12-19   
2          258     2023-11-01     2023-11-07   
3          257     2023-09-20     2023-09-26   
4          256     2023-08-02     2023-08-08   
5          255     2023-06-21     2023-06-27   
6          254     2023-05-03     2023-05-09   
7          253     2023-03-22     2023-03-28   
8          252     2023-02-01     2023-02-07   
9          251     2022-12-07     2022-12-13   
10         250     2022-10-26     2022-11-01   
11         249     2022-09-21     2022-09-27   
12         248     2022-08-03     2022-08-09   
13         247     2022-06-15     2022-06-21   
14         246     2022-05-04     2022-05-10   
15         245     2022-03-16     2022-03-22   
16         244     2022-02-02     2022-02-08   
17         243     2021-12-08     2021-12-14   
18         242     2021-10-27     2021-11-03   
19         241     2021-09-22     2021-09-28   
20         240     2021-08-04     2021-08-10   
21         239     2021-06-16     2021-06-22   
22         238     2021-05-05     2021-05-11   
23         237     2021-03-17     2021-03-23   
24         236     2021-01-20     2021-01-26   
25         235     2020-12-09     2020-12-15   
26         234     2020-10-28     2020-11-03   
27         233     2020-09-16     2020-09-22   
28         232     2020-08-05     2020-08-11   
29         231     2020-06-17     2020-06-23   
30         230     2020-05-06     2020-05-12   
31         229     2020-03-18     2020-03-23   
32         228     2020-02-05     2020-02-11   
33         227     2019-12-11     2019-12-17   
34         226     2019-10-30     2019-11-05   
35         225     2019-09-18     2019-09-24   
36         224     2019-07-31     2019-08-06   
37         223     2019-06-19     2019-06-25   
38         222     2019-05-08     2019-05-14   
39         221     2019-03-20     2019-03-26   
40         220     2019-02-06     2019-02-12   
41         219     2018-12-12     2018-12-18   
42         218     2018-10-31     2018-11-06   
43         217     2018-09-19     2018-09-25   
44         216     2018-08-01     2018-08-07   
45         215     2018-06-20     2018-06-26   
46         214     2018-05-16     2018-05-22   
47         213     2018-03-21     2018-03-27   
48         212     2018-02-07     2018-02-15   
49         211     2017-12-06     2017-12-12   
50         210     2017-10-25     2017-10-31   
51         209     2017-09-06     2017-09-12   
52         208     2017-07-26     2017-08-01   
53         207     2017-05-31     2017-06-06   
54         206     2017-04-12     2017-04-18   
55         205     2017-02-22     2017-03-02   
56         204     2017-01-11     2017-01-17   
57         203     2016-11-30     2016-12-06   
58         202     2016-10-19     2016-10-25   

                                           titulo  
0              260ª Reunião - 30-31 janeiro, 2024  
1             259ª Reunião - 12-13 dezembro, 2023  
2    258ª Reunião - 31 outubro - 1 novembro, 2023  
3             257ª Reunião - 19-20 setembro, 2023  
4                 256ª Reunião - 1-2 agosto, 2023  
5                255ª Reunião - 20-21 junho, 2023  
6                   254ª Reunião - 2-3 maio, 2023  
7                253ª Reunião - 21-22 março, 2023  
8   252ª Reunião - 31 janeiro - 1 fevereiro, 2023  
9               251ª Reunião - 6-7 dezembro, 2022  
10             250ª Reunião - 25-26 outubro, 2022  
11            249ª Reunião - 20-21 setembro, 2022  
12                248ª Reunião - 2-3 agosto, 2022  
13               247ª Reunião - 14-15 junho, 2022  
14                  246ª Reunião - 3-4 maio, 2022  
15               245ª Reunião - 15-16 março, 2022  
16             244ª Reunião - 1-2 fevereiro, 2022  
17              243ª Reunião - 7-8 dezembro, 2021  
18             242ª Reunião - 26-27 outubro, 2021  
19            241ª Reunião - 21-22 setembro, 2021  
20                240ª Reunião 

## Collecting the text of each minute presented in the dataframe 

In [270]:
list_url = "https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detalhes?nro_reuniao=" + df_aux['nroReuniao'].astype(str)

In [271]:
list_url

0     https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detalhes?nro_reuniao=260
1     https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detalhes?nro_reuniao=259
2     https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detalhes?nro_reuniao=258
3     https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detalhes?nro_reuniao=257
4     https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detalhes?nro_reuniao=256
5     https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detalhes?nro_reuniao=255
6     https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detalhes?nro_reuniao=254
7     https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detalhes?nro_reuniao=253
8     https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detalhes?nro_reuniao=252
9     https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detalhes?nro_reuniao=251
10    https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detalhes?nro_reuniao=250
11    https://www.bcb.gov.br/api/servico/sitebcb/copom/atas_detal

Saving all the text from each minute

In [272]:
def collect_ata_data(url):
    """Fetches and processes JSON data from the given URL.

    Args:
        url (str): The URL of the API endpoint.

    Returns:
        dict: The parsed JSON data, if successful.
              None: If there's an error.
    """

    try:
        with urllib.request.urlopen(url) as response:
            if response.status == 200:  # Check for successful response
                data = json.load(response)
                # Process the data here (example: return data)
                return data
            else:
                print(f"Error fetching data from {url}: HTTP status {response.status}")
                return None
    except Exception as e:
        print(f"Error fetching data from {url}: {e}")
        return None

# Create a list of URLs from your `list_url` Series
urls = list_url.tolist()  # Assuming list_url is a pandas Series

# Iterate through the URLs and collect data
ata_data = []
with tqdm(total=len(urls)) as pbar:
    for url in urls:
        data = collect_ata_data(url)
        if data:
            ata_data.append(data)
        pbar.update(1)


100%|██████████| 59/59 [02:15<00:00,  2.30s/it]


Observing the data imported

In [273]:
ata_data

[{'conteudo': [{'nroReuniao': 260,
    'dataReferencia': '2024-01-31',
    'dataPublicacao': '2024-02-06',
    'titulo': '260ª Reunião - 30-31 janeiro, 2024',
    'urlPdfAta': 'https://www.bcb.gov.br/content/copom/atascopom/Copom260-not20240131260.pdf',
    'textoAta': '<div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p class="paragrafo">1. O ambiente externo segue volátil, marcado pelo debate sobre o início do processo de flexibilização da política monetária nas principais economias e por sinais de queda de núcleos de inflação, que ainda permanecem em níveis elevados em diversos países.</p><p class="paragrafo">2. Os bancos centrais das principais economias permanecem determinados em promover a convergência das taxas de inflação para suas metas em um ambiente marcado por pressões nos mercados de trabalho. O Comitê avalia que o cenário segue exigindo cautela por parte de países emergentes.</p><p c

As observed there is no text available until the minute 231. Therefore we'll collect the text from each .pdf file.

In [274]:
# Extraction the needed info 
df = [{'nroReuniao': relatorio['conteudo'][0]['nroReuniao'], 'textoAta': relatorio['conteudo'][0]['textoAta']
      , 'urlPdfAta' : relatorio['conteudo'][0]['urlPdfAta']} for relatorio in ata_data]

# Criating a Dataframe from the dictionary list created above
df = pd.DataFrame(df)

# showing results
df

nroReuniao  \
0          260   
1          259   
2          258   
3          257   
4          256   
5          255   
6          254   
7          253   
8          252   
9          251   
10         250   
11         249   
12         248   
13         247   
14         246   
15         245   
16         244   
17         243   
18         242   
19         241   
20         240   
21         239   
22         238   
23         237   
24         236   
25         235   
26         234   
27         233   
28         232   
29         231   
30         230   
31         229   
32         228   
33         227   
34         226   
35         225   
36         224   
37         223   
38         222   
39         221   
40         220   
41         219   
42         218   
43         217   
44         216   
45         215   
46         214   
47         213   
48         212   
49         211   
50         210   
51         209   
52         208   
53         207   
54         206   
55         205   
56         204   
57         203   
58         202   

                                                                                                                                                 textoAta  \
0   <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
1   <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
2   <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
3   <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
4   <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
5   <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
6   <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
7   <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
8   <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
9   <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
10  <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
11  <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
12  <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
13  <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
14  <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
15  <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário do Copom<sup>1</sup></h3><p clas...   
16  <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura econômica e do cenário de referência do Copom<sup>1</su...   
17  <div id="atacompleta"><div id="ataconteudo"><h3 class="secao">A) Atualização da conjuntura ec

Creating a List with all .pdf files

In [275]:
url_lista = list(df['urlPdfAta'])
url_lista

['https://www.bcb.gov.br/content/copom/atascopom/Copom260-not20240131260.pdf',
 'https://www.bcb.gov.br/content/copom/atascopom/Copom259-not20231213259.pdf',
 'https://www.bcb.gov.br/content/copom/atascopom/Copom258-not20231101258.pdf',
 'https://www.bcb.gov.br/content/copom/atascopom/Copom257-not20230920257.pdf',
 'https://www.bcb.gov.br/content/copom/atascopom/Copom256-not20230802256.pdf',
 'https://www.bcb.gov.br/content/copom/atascopom/Copom255-not20230621255.pdf',
 'https://www.bcb.gov.br/content/copom/atascopom/Copom254-not20230503254.pdf',
 'https://www.bcb.gov.br/content/copom/atascopom/Copom253-not20230322253.pdf',
 'https://www.bcb.gov.br/content/copom/atascopom/Copom252-not20230201252.pdf',
 'https://www.bcb.gov.br/content/copom/atascopom/Copom251-not20221207251.pdf',
 'https://www.bcb.gov.br/content/copom/atascopom/Copom250-not20221026250.pdf',
 'https://www.bcb.gov.br/content/copom/atascopom/Copom249-not20220921249.pdf',
 'https://www.bcb.gov.br/content/copom/atascopom/Cop

Saving all .pdf files for further processing.

In [276]:
def download_pdfs(url_lista):
    file_list = []
    for url in url_lista:
        nome_arquivo = url[-22:-19]
        urllib.request.urlretrieve(url, nome_arquivo + ".pdf") 
        file_list.append(nome_arquivo + ".pdf")
    return file_list


file_list = download_pdfs(url_lista)

In [277]:
print(file_list)

['260.pdf',
 '259.pdf',
 '258.pdf',
 '257.pdf',
 '256.pdf',
 '255.pdf',
 '254.pdf',
 '253.pdf',
 '252.pdf',
 '251.pdf',
 '250.pdf',
 '249.pdf',
 '248.pdf',
 '247.pdf',
 '246.pdf',
 '245.pdf',
 '244.pdf',
 '243.pdf',
 '242.pdf',
 '241.pdf',
 '240.pdf',
 '239.pdf',
 '238.pdf',
 '237.pdf',
 '236.pdf',
 '235.pdf',
 '234.pdf',
 '233.pdf',
 '232.pdf',
 '231.pdf',
 '230.pdf',
 '229.pdf',
 '228.pdf',
 '227.pdf',
 '226.pdf',
 '225.pdf',
 '224.pdf',
 '223.pdf',
 '222.pdf',
 '221.pdf',
 '220.pdf',
 '219.pdf',
 '218.pdf',
 '217.pdf',
 '216.pdf',
 '215.pdf',
 '214.pdf',
 '213.pdf',
 '212.pdf',
 '211.pdf',
 '210.pdf',
 '209.pdf',
 '208.pdf',
 '207.pdf',
 '206.pdf',
 '205.pdf',
 '204.pdf',
 '203.pdf',
 '202.pdf']

# Data processing/ Creating Variables

Opening files, and converting text to .txt 

In [278]:
def pdf_to_txt(file_list):
    for arquivo in file_list:
        with fitz.open(arquivo) as pdf:
            texto = ""
            
            for pagina_num in range(2, pdf.page_count):
                pagina = pdf[pagina_num]
                texto += pagina.get_text()
            
            with open(arquivo.replace('.pdf', '.txt'), 'w', encoding='utf-8') as f:
                f.write(texto)

In [279]:
pdf_to_txt(file_list)

# Collecting and processing Data / Creating Variables

Opening .txt files

In [280]:
# Obtém o caminho do diretório atual
current_directory = os.getcwd()

# Lista todos os arquivos no diretório atual
all_files = os.listdir(current_directory)

# Filtra apenas os arquivos com extensão .txt
file_list = [file for file in all_files if file.endswith(".txt")]

# Exibe a lista de nomes de arquivos .txt
print(file_list)


['202.txt', '203.txt', '204.txt', '205.txt', '206.txt', '207.txt', '208.txt', '209.txt', '210.txt', '211.txt', '212.txt', '213.txt', '214.txt', '215.txt', '216.txt', '217.txt', '218.txt', '219.txt', '220.txt', '221.txt', '222.txt', '223.txt', '224.txt', '225.txt', '226.txt', '227.txt', '228.txt', '229.txt', '230.txt', '231.txt', '232.txt', '233.txt', '234.txt', '235.txt', '236.txt', '237.txt', '238.txt', '239.txt', '240.txt', '241.txt', '242.txt', '243.txt', '244.txt', '245.txt', '246.txt', '247.txt', '248.txt', '249.txt', '250.txt', '251.txt', '252.txt', '253.txt', '254.txt', '255.txt', '256.txt', '257.txt', '258.txt', '259.txt', '260.txt']


Using stopwords in the language of the text (portuguese).

In [281]:
# Load stopwords
stopwords = nltk.corpus.stopwords.words('portuguese')

In [282]:
# Additional list of words to be removed
ad_list = ['meses', 'ano', 'mês', 'relação', 'doze', 'período', 'sobre', 'janeiro', 'fevereiro', 'março', 'abril',
           'maio', 'junho', 'julho', 'agosto', 'setembro', 'outubro', 'novembro', 'dezembro', 'nível', '-', 'bilhões',
           'anterior', 'copom', 'banco', 'central', 'após', 'desde', 'comparação', 'respectivamente', 'anterior',
           'monetária', 'índice', 'ante', 'pp', 'variação', 'reunião', 'trimestre', 'comitê', 'chefe', 'segue',
           'departamento', 'membros', 'ainda', 'cenário', '2022', '2023', 'a.a.', '2023.', 'básico']

In [283]:
stopwords = set(list(stopwords) + list(string.punctuation) + ad_list)

In [284]:
dict_top = {}

for arquivo in file_list:
    with open(arquivo.replace('.pdf', '.txt'),'r', encoding='utf-8') as f:
        # Remover caracteres não alfanuméricos
        texto = f.read().replace(r'[^\w\s\-áéíóúãõâêîôûàèìòùäëïöüç]', '').lower()
        # Substituir quebras de linha por espaços
        texto = texto.replace('\n', ' ')
        frequency_distribution = FreqDist(word.lower() for word in word_tokenize(texto) if word not in stopwords)
        dict_top[arquivo.replace('.pdf', '')] = frequency_distribution.most_common(10)

Creating a Dataframe with the 10 most common words in each minute.

In [285]:
df_words = pd.DataFrame(dict_top)

In [286]:
df_words = df_words.T

# Obtém os nomes atuais das colunas
nomes_colunas_atuais = df_words.columns.tolist()

# Gera os novos nomes de colunas com o prefixo "df_words"
novos_nomes_colunas = ['df_words_' + str(coluna) for coluna in nomes_colunas_atuais]

# Renomeia as colunas
df_words = df_words.rename(columns=dict(zip(nomes_colunas_atuais, novos_nomes_colunas)))



Creating a Dataframe with the 10 most common bigrams in each minute.

In [287]:
dict_top_bigram = {}

for arquivo in file_list:
    with open(arquivo.replace('.pdf', '.txt'), 'r', encoding='utf-8') as f:
        # Remover caracteres não alfanuméricos
        texto = f.read().replace(r'[^\w\s\-áéíóúãõâêîôûàèìòùäëïöüç]', '').lower()
        # Substituir quebras de linha por espaços
        texto = texto.replace('\n', ' ')
        frequency_distribution = FreqDist(word for word in nltk.bigrams([word.lower() for word in word_tokenize(texto) if word not in stopwords]))
        dict_top_bigram[arquivo.replace('.pdf', '')] = frequency_distribution.most_common(10)

In [288]:
df_bigram = pd.DataFrame(dict_top_bigram)

In [289]:
df_bigram = df_bigram.T

# Obtém os nomes atuais das colunas
nomes_colunas_atuais = df_bigram.columns.tolist()

# Gera os novos nomes de colunas com o prefixo "df_words"
novos_nomes_colunas = ['df_bigram_' + str(coluna) for coluna in nomes_colunas_atuais]

# Renomeia as colunas
df_bigram = df_bigram.rename(columns=dict(zip(nomes_colunas_atuais, novos_nomes_colunas)))

Creating a Dataframe with the 10 most common trigrams in each minute.

In [290]:
dict_top_trigram = {}

for arquivo in file_list:
    with open(arquivo.replace('.pdf', '.txt'), 'r', encoding='utf-8') as f:
        # Remover caracteres não alfanuméricos
        texto = f.read().replace(r'[^\w\s\-áéíóúãõâêîôûàèìòùäëïöüç]', '').lower()
        # Substituir quebras de linha por espaços
        texto = texto.replace('\n', ' ')
        frequency_distribution = FreqDist(word for word in nltk.trigrams([word.lower() for word in word_tokenize(texto) if word not in stopwords]))
        dict_top_trigram[arquivo.replace('.pdf', '')] = frequency_distribution.most_common(10)

In [291]:
df_trigram = pd.DataFrame(dict_top_trigram)

In [292]:
df_trigram = df_trigram.T

# Obtém os nomes atuais das colunas
nomes_colunas_atuais = df_trigram.columns.tolist()

# Gera os novos nomes de colunas com o prefixo "df_words"
novos_nomes_colunas = ['df_trigram_' + str(coluna) for coluna in nomes_colunas_atuais]

# Renomeia as colunas
df_trigram = df_trigram.rename(columns=dict(zip(nomes_colunas_atuais, novos_nomes_colunas)))

Collecting our target: the communication expressing the decision of the Central Bank of Brazil regarding the Selic Interest Rate (Taxa Selic).

In [293]:
# Using 1000 to ensure all available data is being collected
url_alvo = 'https://www.bcb.gov.br/api/servico/sitebcb/copom/comunicados?quantidade=1000'

In [294]:
with urllib.request.urlopen(url_alvo) as url:
    data = json.load(url)
    print(data)

{'conteudo': [{'nro_reuniao': 260, 'dataReferencia': '2024-01-31', 'titulo': '260ª reunião - Copom reduz a taxa Selic para 11,25% a.a.'}, {'nro_reuniao': 259, 'dataReferencia': '2023-12-13', 'titulo': '259ª reunião - Copom reduz a taxa Selic para 11,75% a.a.'}, {'nro_reuniao': 258, 'dataReferencia': '2023-11-01', 'titulo': '258ª reunião - Copom reduz a taxa Selic para 12,25% a.a.'}, {'nro_reuniao': 257, 'dataReferencia': '2023-09-20', 'titulo': '257ª reunião - Copom reduz a taxa Selic para 12,75% a.a.'}, {'nro_reuniao': 256, 'dataReferencia': '2023-08-02', 'titulo': '256ª reunião - Copom reduz a taxa Selic para 13,25% a.a.'}, {'nro_reuniao': 255, 'dataReferencia': '2023-06-21', 'titulo': '255ª reunião - Copom mantém a taxa Selic em 13,75% a.a.'}, {'nro_reuniao': 254, 'dataReferencia': '2023-05-03', 'titulo': '254ª reunião - Copom mantém a taxa Selic em 13,75% a.a.'}, {'nro_reuniao': 253, 'dataReferencia': '2023-03-22', 'titulo': '253ª reunião - Copom mantém a taxa Selic em 13,75% a.a.'

In [295]:
df_alvo = pd.DataFrame(data['conteudo'])

In [296]:
index_202 = df_alvo[df_alvo['nro_reuniao'] == 202].index[0]
df_alvo = df_alvo.loc[:index_202]

Simplifying the output filtering only if the interst rate : reduces, maintains or increases

In [297]:
# Assuming df_alvo is your DataFrame
df_alvo['alvo'] = df_alvo['titulo'].apply(lambda x: 'reduz' if 'reduz' in x else ('mantém' if 'mantém' in x else 'eleva'))

Merging all data in one Dataframe

In [298]:
df_words.index = df_words.index.str.rstrip('.txt')
df_bigram.index = df_bigram.index.str.rstrip('.txt')
df_trigram.index = df_trigram.index.str.rstrip('.txt')


In [299]:
# Converta o índice de df_words para o mesmo tipo da coluna 'nro_reuniao' de df_alvo
df_words.index = df_words.index.astype(df_alvo['nro_reuniao'].dtype)
df_bigram.index = df_bigram.index.astype(df_alvo['nro_reuniao'].dtype)
df_trigram.index = df_trigram.index.astype(df_alvo['nro_reuniao'].dtype)

# Realize o inner join
df = pd.merge(df_alvo, df_words, left_on='nro_reuniao', right_index=True, how='inner')
df = pd.merge(df, df_bigram, left_on='nro_reuniao', right_index=True, how='inner')
df = pd.merge(df, df_trigram, left_on='nro_reuniao', right_index=True, how='inner')

In [300]:
df

nro_reuniao dataReferencia  \
0           260     2024-01-31   
1           259     2023-12-13   
2           258     2023-11-01   
3           257     2023-09-20   
4           256     2023-08-02   
5           255     2023-06-21   
6           254     2023-05-03   
7           253     2023-03-22   
8           252     2023-02-01   
9           251     2022-12-07   
10          250     2022-10-26   
11          249     2022-09-21   
12          248     2022-08-03   
13          247     2022-06-15   
14          246     2022-05-04   
15          245     2022-03-16   
16          244     2022-02-02   
17          243     2021-12-08   
18          242     2021-10-27   
19          241     2021-09-22   
20          240     2021-08-04   
21          239     2021-06-16   
22          238     2021-05-05   
23          237     2021-03-17   
24          236     2021-01-20   
25          235     2020-12-09   
26          234     2020-10-28   
27          233     2020-09-16   
28          232     2020-08-05   
29          231     2020-06-17   
30          230     2020-05-06   
31          229     2020-03-18   
32          228     2020-02-05   
33          227     2019-12-11   
34          226     2019-10-30   
35          225     2019-09-18   
36          224     2019-07-31   
37          223     2019-06-19   
38          222     2019-05-08   
39          221     2019-03-20   
40          220     2019-02-06   
41          219     2018-12-12   
42          218     2018-10-31   
43          217     2018-09-19   
44          216     2018-08-01   
45          215     2018-06-20   
46          214     2018-05-16   
47          213     2018-03-21   
48          212     2018-02-07   
49          211     2017-12-06   
50          210     2017-10-25   
51          209     2017-09-06   
52          208     2017-07-26   
53          207     2017-05-31   
54          206     2017-04-12   
55          205     2017-02-22   
56          204     2017-01-11   
57          203     2016-11-30   
58          202     2016-10-19   

                                                          titulo    alvo  \
0       260ª reunião - Copom reduz a taxa Selic para 11,25% a.a.   reduz   
1       259ª reunião - Copom reduz a taxa Selic para 11,75% a.a.   reduz   
2       258ª reunião - Copom reduz a taxa Selic para 12,25% a.a.   reduz   
3       257ª reunião - Copom reduz a taxa Selic para 12,75% a.a.   reduz   
4       256ª reunião - Copom reduz a taxa Selic para 13,25% a.a.   reduz   
5        255ª reunião - Copom mantém a taxa Selic em 13,75% a.a.  mantém   
6        254ª reunião - Copom mantém a taxa Selic em 13,75% a.a.  mantém   
7        253ª reunião - Copom mantém a taxa Selic em 13,75% a.a.  mantém   
8        252ª reunião - Copom mantém a taxa Selic em 13,75% a.a.  mantém   
9        251ª reunião - Copom mantém a taxa Selic em 13,75% a.a.  mantém   
10       250ª reunião - Copom mantém a taxa Selic em 13,75% a.a.  mantém   
11       249ª reunião - Copom mantém a taxa Selic em 13,75% a.a.  mantém   
12      248ª reunião - Copom eleva a taxa Selic para 13,75% a.a.   eleva   
13      247ª reunião - Copom eleva a taxa Selic para 13,25% a.a.   eleva   
14      246ª reunião - Copom eleva a taxa Selic para 12,75% a.a.   eleva   
15      245ª reunião - Copom eleva a taxa Selic para 11,75% a.a.   eleva   
16      244ª reunião - Copom eleva a taxa Selic para 10,75% a.a.   eleva   
17       243ª reunião - Copom eleva a taxa Selic para 9,25% a.a.   eleva   
18       242ª reunião - Copom eleva a taxa Selic para 7,75% a.a.   eleva   
19       241ª reunião - Copom eleva a taxa Selic para 6,25% a.a.   eleva   
20       240ª reunião - Copom eleva a taxa Selic para 5,25% a.a.   eleva   
21       239ª reunião - Copom eleva a taxa Selic para 4,25% a.a.   eleva   
22       238ª reunião - Copom eleva a taxa Selic para 3,50% a.a.   eleva   
23       237ª reunião - Copom eleva a taxa Selic para 2,75% a.a.   eleva   
24        236ª reunião - Copom mantém a taxa Selic em 2,00% a.a

Cleaning Dataframe

In [301]:
# List of columns to drop
columns_to_drop = ['nro_reuniao', 'dataReferencia', 'titulo']

# Separate features (X) and target variable (y)
df_ml = df.drop(columns=columns_to_drop, axis=1)

Shifting the target by 1 cell. 

The objective here is to predict the next output based on the current minute text. Therefore, we need the last minute to predict the next decision.

In [302]:
df_ml['alvo'] = df_ml['alvo'].shift(1)

Making copies of the DataFrame for further operations

In [303]:
df_pred = df_ml.copy()

Converting the Target into 0,1 or 2 values

# Applying Machine Learning Models to predict the Output of the Brazil Central Bank decision

In [304]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.tokenize import word_tokenize
import pickle

# Carregar o modelo treinado
with open('model_Selic_Rate.pkl', 'rb') as model_file:
    model = pickle.load(model_file)

# Carregar o vetorizador TF-IDF usado durante o treinamento
with open('vectorizer_Selic_Rate.pkl', 'rb') as vectorizer_file:
    vectorizer = pickle.load(vectorizer_file)

# Concatenação dos dados textuais
df_text = df_pred.apply(lambda row: ' '.join(map(str, row)), axis=1)

# Vetorização do texto usando TF-IDF
X_df = vectorizer.transform(df_text)

# Fazer previsões para todo o DataFrame
predictions_df = model.predict(X_df)
probabilities_df = model.predict_proba(X_df)


# Adicionando as previsões ao DataFrame original
df_pred['predictions'] = predictions_df
# Adicionando as probabilidades ao DataFrame original para cada classe
for i, class_label in enumerate(model.classes_):
    col_name = f'probability_{class_label}'
    df_pred[col_name] = probabilities_df[:, i]

# Exibindo o DataFrame resultante
df_pred

alvo      df_words_0          df_words_1          df_words_2  \
0      NaN  (inflação, 38)      (política, 22)      (dinâmica, 17)   
1    reduz  (inflação, 34)      (política, 22)  (expectativas, 13)   
2    reduz  (inflação, 40)      (política, 21)     (atividade, 14)   
3    reduz  (inflação, 48)      (política, 21)      (dinâmica, 16)   
4    reduz  (inflação, 58)      (política, 21)  (expectativas, 21)   
5    reduz  (inflação, 44)  (expectativas, 23)      (processo, 20)   
6   mantém  (inflação, 47)      (política, 21)  (expectativas, 20)   
7   mantém  (inflação, 49)      (política, 26)  (expectativas, 22)   
8   mantém  (inflação, 42)      (política, 24)  (expectativas, 19)   
9   mantém  (inflação, 37)      (política, 17)  (expectativas, 13)   
10  mantém  (inflação, 28)      (política, 15)  (expectativas, 13)   
11  mantém  (inflação, 35)      (política, 20)  (expectativas, 15)   
12  mantém  (inflação, 31)      (política, 19)        (riscos, 11)   
13   eleva  (inflação, 27)      (política, 17)        (riscos, 12)   
14   eleva  (inflação, 25)      (política, 13)        (riscos, 11)   
15   eleva  (inflação, 30)     (projeções, 14)      (cenários, 12)   
16   eleva  (inflação, 26)    (referência, 12)         (risco, 10)   
17   eleva  (inflação, 23)   (expectativas, 9)      (projeções, 9)   
18   eleva  (inflação, 26)        (riscos, 11)       (balanço, 10)   
19   eleva  (inflação, 21)        (riscos, 11)      (política, 10)   
20   eleva  (inflação, 31)      (política, 14)        (riscos, 12)   
21   eleva  (inflação, 31)        (riscos, 13)     (projeções, 11)   
22   eleva  (inflação, 23)        (riscos, 12)      (atividade, 9)   
23   eleva  (inflação, 26)        (riscos, 12)          (meta, 11)   
24   eleva  (inflação, 26)     (econômica, 10)        (riscos, 10)   
25  mantém  (inflação, 32)          (2021, 11)      (política, 10)   
26  mantém  (inflação, 24)          (taxa, 13)         (juros, 13)   
27  mantém  (inflação, 30)         (juros, 15)          (taxa, 13)   
28  mantém  (inflação, 29)         (juros, 16)          (taxa, 13)   
29   reduz  (inflação, 16)      (pandemia, 13)      (economia, 10)   
30   reduz  (inflação, 15)          (taxa, 13)         (juros, 12)   
31   reduz  (inflação, 13)          (taxa, 13)      (política, 12)   
32   reduz  (inflação, 20)      (política, 19)      (economia, 15)   
33   reduz  (inflação, 18)      (economia, 14)      (política, 14)   
34   reduz  (inflação, 31)      (economia, 24)      (política, 21)   
35   reduz  (inflação, 38)          (taxa, 23)      (economia, 22)   
36   reduz      (taxa, 31)      (inflação, 31)      (economia, 29)   
37   reduz  (inflação, 34)      (economia, 25)     (econômica, 19)   
38  mantém  (inflação, 32)      (economia, 26)          (taxa, 16)   
39  mantém  (inflação, 32)      (economia, 26)      (política, 17)   
40  mantém  (inflação, 30)      (economia, 22)      (política, 21)   
41  mantém  (inflação, 30)      (política, 19)      (economia, 18)   
42  mantém  (inflação, 45)      (política, 26)      (economia, 20)   
43  mantém  (inflação, 46)      (política, 26)      (economia, 22)   
44  mantém  (inflação, 48)      (economia, 23)      (política, 23)   
45  mantém  (inflação, 43)      (economia, 23)      (política, 20)   
46  mantém  (inflação, 52)      (política, 26)      (economia, 23)   
47  mantém  (inflação, 40)      (política, 25)      (economia, 21)   
48   reduz  (inflação, 38)      (política, 20)      (economia, 18)   
49   reduz  (inflação, 33)      (política, 21)      (economia, 18)   
50   reduz  (inflação, 29)      (política, 18)      (economia, 17)   
51   reduz  (inflação, 34)      (economia, 20)      (política, 17)   
52   reduz  (inflação, 25)      (economia, 24)      (política, 16)   
53   reduz  (inflação, 28)      (economia, 22)     (incerteza, 16)   
54   reduz  (inflação, 26)      (economia, 22)     (econômica, 15)   
55   reduz  (inflação, 26)      (economia, 24)      (política, 17)   
56   reduz  (inf

In [305]:
# Codificação do alvo
class_mapping = {0 : 'reduces', 1 : 'maintains', 2 : 'increases'}
df_pred['predictions'] = df_pred['predictions'].map(class_mapping)

In [306]:
df_pred.rename(columns={'probability_0': 'reduces', 'probability_1': 'maintains', 'probability_2': 'increases'}, inplace=True)

# Showing Prediction for Next Meeting

In [307]:
# Delete all filles created during the data processing

def deletar_arquivos_com_extensao(pasta, extensoes):
    for extensao in extensoes:
        padrao = os.path.join(pasta, f'*.{extensao}')
        arquivos = glob.glob(padrao)
        for arquivo in arquivos:
            try:
                os.remove(arquivo)
                print(f'Arquivo {arquivo} removido com sucesso.')
            except Exception as e:
                print(f'Erro ao remover {arquivo}: {str(e)}')

# Obtém o diretório atual do script
diretorio_atual = os.getcwd()

# Especifica as extensões dos arquivos a serem excluídos
extensoes_para_excluir = ['txt', 'pdf']

# Chama a função para deletar os arquivos com as extensões especificadas
deletar_arquivos_com_extensao(diretorio_atual, extensoes_para_excluir)

Arquivo c:\Users\lvbru\OneDrive\Documentos\Bruno\Ciencia_de_Dados\Portifolio\NLP-Brazil_Central_Bank_Interest_Rate_SELIC\NLP-Brazil_Central_Bank_Interest_Rate_SELIC\202.txt removido com sucesso.
Arquivo c:\Users\lvbru\OneDrive\Documentos\Bruno\Ciencia_de_Dados\Portifolio\NLP-Brazil_Central_Bank_Interest_Rate_SELIC\NLP-Brazil_Central_Bank_Interest_Rate_SELIC\203.txt removido com sucesso.
Arquivo c:\Users\lvbru\OneDrive\Documentos\Bruno\Ciencia_de_Dados\Portifolio\NLP-Brazil_Central_Bank_Interest_Rate_SELIC\NLP-Brazil_Central_Bank_Interest_Rate_SELIC\204.txt removido com sucesso.
Arquivo c:\Users\lvbru\OneDrive\Documentos\Bruno\Ciencia_de_Dados\Portifolio\NLP-Brazil_Central_Bank_Interest_Rate_SELIC\NLP-Brazil_Central_Bank_Interest_Rate_SELIC\205.txt removido com sucesso.
Arquivo c:\Users\lvbru\OneDrive\Documentos\Bruno\Ciencia_de_Dados\Portifolio\NLP-Brazil_Central_Bank_Interest_Rate_SELIC\NLP-Brazil_Central_Bank_Interest_Rate_SELIC\206.txt removido com sucesso.
Arquivo c:\Users\lvbru\On

In [308]:
print('-----------------------------------------------------------------------------------------')
print(f"Based on the text of the last minute from Brazil's Central Bank, minute number "+str(df['nro_reuniao'][0]))
print(f"the next resolution, concering the Brazilian interest ratio, will be:")
print(f' "' + str(df_pred['predictions'][0]) + '" ')
print('-----------------------------------------------------------------------------------------')
print(f'the probabilities of the predictions are:')
print(df_pred[['reduces','maintains','increases']].iloc[0])
print('-----------------------------------------------------------------------------------------')

-----------------------------------------------------------------------------------------
Based on the text of the last minute from Brazil's Central Bank, minute number 260
the next resolution, concering the Brazilian interest ratio, will be:
 "reduces" 
-----------------------------------------------------------------------------------------
the probabilities of the predictions are:
reduces      0.69
maintains    0.18
increases    0.13
Name: 0, dtype: float64
-----------------------------------------------------------------------------------------
